### ISSUE TITLE
Problems with current Ruby 2.2 Dependency

### ISSUE THREAD

**Issue by bastelfreak** [guard/listen#374]
Hi,

listen 3.0.6 needed at least Ruby1.9.3, you bumped the Ruby version to at least 2.2 in listen3.1.0. This is a major change in my opinion and will break many current setups, and is probably not compatible with semantic versioning. Could you please revoke the latest release or release a new 3.X that has no dependency on a newer ruby?

Please don't get me wrong, I really really love new software. But please require ruby2.2 after a major bump of listen, in a 4.X release.

**Comment by e2** [comment]
Thanks for the opinion on this.

Understand it was a tough choice, so if you want me to revert, help me by making very strong and specific justifications first.

### 1. It's a RubyGems bug/problem/issue, not a Listen one:

The version is clearly stated in the gemspec, just like any other dependency:

https://github.com/guard/listen/blob/master/listen.gemspec#L25

So if you can even **install** listen on a different Ruby version, that's a RubyGems problem. (?)


### 2. Security

This one is big. 1.9.3 is long dead and not supported. 2.0 is EOL. 2.1 has only security patches and it won't be long until it's EOL to. 2.2 is now recommended on ruby-lang.org. 2.3 has many important features is almost ready for production (with the exception of one bugfix that hasn't been released yet).

Security should be a no-brainer. Even if migrating was hard, it's still too important. It's priority #1, especially since people use Listen on servers and company workstations.

### 3. It's SemVer ok

I considered a major version bump (4.x), but I couldn't find this practice justified by SemVer. (Please point me to a line/discussion where it's mandatory to bump to 4.x). 

Listen's public API didn't change. If it did, I would have to bump it to 4.x. But it didn't change. What changed was a version of a dependency. If you depend on Ruby 1.9.2, that should be explicit in your project/gem.

### 4. Upgrading is usually painless.

Please point me to the exact problems if that's not the case. 

If installing a newer Ruby is a problem, then it's a packaging problem - and not a Listen problem. Expecting Listen to support a previous Ruby all because it's "hard to install" is not an issue I should make sacrifices to deal with.

I do hear about "problems", but no specifics, ever. Never saw one in real life.

### 5. You don't have to upgrade

It's a minor version bump, so there's room for me to accept backporting PRs for Listen 3.0.x.

So you can lock to Listen '~> 3.0.x'. You should always read the upgrade notes "just in case" before upgrading - even just because people make mistakes. Or, there might be new recommendations for security issues.

### 6. To avoid Ruby becoming dead

No one is ever excited about supporting Ruby 1.8.7. Expecting it to be supported today is the most detrimental thing you can do to the Ruby community.

It's similar to expecting support for Ruby 1.9.3, which is downright unreasonable, given there's 2.0, 2.1, 2.2, and 2.3 out there. 2.x has been around for a long, long, long, long, loooooooooong time.

How much more time to people need to upgrade? The answer is: if you give people an eternity to upgrade, they will happily wait for an eternity to upgrade.

It's a drain on everyone and no one gets paid for having their excitement "with new stuff" killed like that. And since open source is free (unpaid) effort, excitement is too important. (Unless someone wants to pay).

This means contributors (people contributing fixes and new features) are more important than ever. Often time, more important than users even (in the long run).

If Ruby 2.3.0 has been released  (after the Ruby team put in SO MUCH INSANE EFFORT AND TIME) and people expect many gems to still be compatible with 1.8.7 even ...

... how is that beneficial to anyone?

A conspiracy theorist will assume it's a great way to kill Ruby. Because you can't use new features. Which basically makes the language dead.

If you notice, almost all gems out there are 99.9999% percent still compatible with Ruby 1.8.7. With maybe the exceptions of the JSON hash syntax, using a few convenience methods, relying on hash sorting.

Almost no one is ACTUALLY using Ruby 2.x features out there!

And the people putting on pressure to be "backward compatible" are not contributing ANYTHING, really!

### 7. Honoring the effort of the Ruby community

Yes, businesses and users expect things to run without changes.

But what is the "reward" for people working hard to constantly keep things improving AND stay insanely backwards compatible?

That's like 10x the effort with no reward.

So I'm asking: what is the reward?

Well, there is none.

It's actually better to "experiment" with Ruby 2.3 than to rely on a "stable" (but dead) 1.9.3. Because there's no point in making 1.9.3 "more stable" and letting 2.3 "stay unstable" for longer.

It's throwing away people's effort. Just because everyone is scared of change.

### 8. No "real" problems to justify the costs.

People complaining quote SemVer (without really having read how it applies in this scenario).

But they don't even post the "issues" they're having.

Because there really aren't any issue, just misdirected "complaints".

E.g. I can exactly point you to a bug that occurs in Ruby 2.3 that hasn't been released yet (but is scheduled to be released soon). Otherwise, I would've switched to Ruby 2.3 due to an important feature.

**Comment by jyaworski** [comment]
I would argue that this is a major release. Jumping from 1.9.3 to 2.2 is not something that those using this in operations or elsewise (that cannot upgrade for whatever reason) can easily deal with. RHEL 7 shipped with 2.0.0, and that was recently released (and will be using for another eight years minimum).

I'm amenable to the argument that we should drop old version support (we dropped 1.8.7 support even though RHEL6 is still very much a thing) but it should be obvious to the users that we have done so.

**Comment by e2** [comment]
> I would argue that this is a major release.

No public API change here. And Listen isn't for "testing Ruby". Does this API change cause problems with using Listen? It doesn't. It only forces people to opt-in into a recommended version of Ruby.

Actually, a major release would suggest a major API overhaul - which would be VERY misleading to users. Just imaging calling Rails 4.3 "Rails 5" because of a bumped dependency on e.g. the 'pg' gem.

Expecting a major release here is placing expectations beyond the scope of a single gem like Listen.

> Jumping from 1.9.3 to 2.2 is not something that those using this in operations or elsewise (that cannot upgrade for whatever reason) can easily deal with.

Either that's a packaging problem (that can be solved by e.g. RVM within a single user account even), or I don't see any specific problem.

Impact: none.

> I'm amenable to the argument that we should drop old version support (we dropped 1.8.7 support even though RHEL6 is still very much a thing) but it should be obvious to the users that we have done so.

Which is exactly what I have done - you can't install the current version of the gem on an older version of Ruby.

It breaks provisioning scripts - which is perfect. The bad outcome would be breaking services running on production servers.

Devops have rollback functionality for failed provisioning, right?

> RHEL 7 shipped with 2.0.0, and that was recently released (and will be using for another eight years minimum).

Appeal to them! They are suggesting they'll be supporting a version of Ruby that already isn't officially supported. Does that mean they aren't going to provide security updates for 8 years?

Not to mention: 2.0.0 is probably the most "experimental" and "outdated" version of Ruby possible at this point. (Even if you take patch level into account).

I'm sure you get the irony and hypocrisy here.

What I mean is: I get the "intention" of using warning users about massive changes with major version bumps. But there's simply nothing to actually warn users about in this case. 

In devops, even a patch-level change can cause major breakage - you can neither "fix" that nor "prevent" that by bumping a version number.

So the question remains: what exactly "broke" and how is Listen's version number the **root cause**?

**Comment by bastelfreak** [comment]
Hi,

first of all: thank you so much for your detailed responses, I really didn't expect that (also on a sunday) :+1: .

---

For my particular issue: I'm managing, with many others, a huge amount of puppet modules. They have rspec tests, at least one of them uses the listen gem as a dependency. An example for a failed run is [here](https://travis-ci.org/voxpupuli/puppet-collectd/builds/125300238).

There is an old open issue on [semver](https://github.com/mojombo/semver/issues/148) about the needed version update after a major dependency update. In my opinion this isn't well-defined and the opinions differ. 

The update from listen3.0.6 to 3.1 will break the system if you're on an older ruby version because it fails to install (ok maybe not completely break but you get into some trouble). Yes this is more of a rubygems issue. Nevertheless, speaking from an Ops perspective: a major update of any software is always itchy, and in the case of ruby not always possible. The ruby bump in the listen gem may be correct in the sense of semver, but it will have a huge impact on productive systems.  This is a thing that most people don't expect on a minor change.

I'm a huge friend of new and up2date systems, I encourage everyone, administrators and developers to drop support for older versions, but please not with a minor update.

---

I think you can close this issue. Just if you want to do me a favour: make a new 4.X release which requires ruby2.2 or even 2.3, and make one last 3.X release which works with ruby1.9

**Comment by e2** [comment]
@bastelfreak - you didn't fully read my reply. :(

1. I posted a link to your failed run (as an update to my comment - if you only read a GitHub email notification from me, you didn't get it)

2. I know about the puppet modules - that's how I found the failing build log. 

3. The error is "works as expected". So it is not a RubyGems issue. RubyGems is working correctly by refusing the install the gem.

4. Yes, I read that semver issue previously. There are cases where this is complex. But here, a shift from 1.9.2 to 2.2 (or 2.3 even) is MUCH smaller than from Ruby 1.8.7 to 1.9.2. (Ask Matz if you're in doubt). Originally, Ruby 1.9 was technically 2.0, but it was named 1.9 to promote migration.

To avoid exactly the problem you brought up here: avoiding to upgrade based on version number instead of based on actual changes and impact.

5. The following is incorrect: 

> The update from listen3.0.6 to 3.1 will break the system if you're on an older ruby version because it fails to install (ok maybe not completely break but you get into some trouble). 

No, it won't "break the system" (a vague term), it will only break PROVISIONING. And this is devops responsibility. If provisioning breaks, that's great, because the services are protected due to a rollback.

And provisioning for devops is "business as usual". The point of devops is to isolate such failures as part of creating a deployment plan.

6. Also, ...

> Nevertheless, speaking from an Ops perspective: a major update of any software is always itchy

"itchy" is not a specific/technical term compared to the points I gave above.

> and in the case of ruby not always possible. 

Packaging problem. That's devops responsibility (or the provider of the distro/OS). Not an internal dependency problem that should be fixed here.

> The ruby bump in the listen gem may be correct in the sense of semver, but it will have a huge impact on productive systems. 

I have yet to hear about ANY specific compelling example that actually occurred. (Or that could occur). And when I say **specific**, I mean that.

> This is a thing that most people don't expect on a minor change.

I "expect" people to read Release notes and changelogs before upgrading software. And, I have yet to see an example of what "harm" I've done to people who upgraded without reading.

Also, it **was** a minor change (based on impact). 

> Just if you want to do me a favour: make a new 4.X release which requires ruby2.2 or even 2.3, and make one last 3.X release which works with ruby1.9

You didn't read my reasons for doing what I did. And what's more important - you STILL didn't give me a compelling example of how I broke anything, really.

**Comment by rymai** [comment]
I agree with @e2, the error you get is valid, and useful!

If you're using a `Gemfile.lock` file that shouldn't be a problem since you won't be able to update the `Gemfile.lock` with Listen 3.1.0 on a system that don't have Ruby 2.2 installed. Apocalypse avoided! ;)

**Comment by e2** [comment]
For what it's worth, RuboCop already works differently depending on Ruby version: https://github.com/guard/listen/blob/master/.rubocop.yml#L11

So supporting multiple versions of Ruby really makes less and less sense as you get deeper into things.

**Comment by rnelson0** [comment]
@e2 You could argue that you did not make public API changes, though that's only listed as a [MUST](http://semver.org/#spec-item-8) reason, no as the only reason to increase MAJOR. Semver also states that it is only permissable to increase the `MINOR version when you add functionality in a backwards-compatible manner`.

However, your public API in the form of [the gemspec](https://github.com/guard/listen/blob/master/listen.gemspec#L25) specified that only ruby 1.9.3 was required until 3 days ago. Unfortunately, semver is [ambiguous](http://semver.org/#if-even-the-tiniest-backwards-incompatible-changes-to-the-public-api-require-a-major-version-bump-wont-i-end-up-at-version-4200-very-rapidly) [about this](http://semver.org/#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api). One place it is not ambiguous is that you did not [issue a minor release with the deprecation in place](http://semver.org/#how-should-i-handle-deprecating-functionality) that I can observe.

I think the responsible path forward is to release a 3.0.next version with the deprecation warning, then start on 4.0.0 using ruby 2.2-isms. Both you and those using `~> 3.0` are able to rely on SemVer expectations.


(Side note: There's some ~100 open issues on the SemVer issue tracker because it's not perfect and doesn't capture all the nuance possible. In a devops world, we should move fast and break things - including semver - but always retain our empathy for our users. It seems to make more sense to stick to the spirit of SemVer while catering to more users than to rigidly align with the text on a webpage and alienate some users, especially when the cost is so low.)

**Comment by e2** [comment]
@rnelson0 - Thanks for the detailed reply and understanding!

Let me address the points for you:

> Semver also states that it is only permissable to increase the MINOR version when you add functionality in a backwards-compatible manner.

How is Ruby 2.3 (with a minor version bump) not backward compatible with Ruby 2.0? (1.9.3 doesn't count, because it's EOL - people had way too long to upgrade already).

> One place it is not ambiguous is that you did not issue a minor release with the deprecation in place that I can observe.

Because I wasn't sure what was being deprecated (specifically). Most of Listen has a Ruby 1.8.7 codebase technically, which Ruby 2.3 accepts pretty much without hiccups (except for hash style, basically).

> I think the responsible path forward is to release a 3.0.next version with the deprecation warning, then start on 4.0.0 using ruby 2.2-isms.

Deprecate what exactly? "Ruby version" isn't an API or a functionality. It isn't an API I control within Listen.

There is no code that needs to be changed in how Listen is used to warrant a major version change.

And I didn't find that beneficial for the community. SemVer isn't more important than the experience of users and contributors.

Also, it's hypocrisy to avoid reading changelogs by relying on SemVer ... yet ... not do the same when it comes to Ruby versions.

> There's some ~100 open issues on the SemVer issue tracker because it's not perfect and doesn't capture all the nuance possible. 

That's why I'm thinking here. I'm going beyond "is this breaking SemVer" and I'm focusing on the Ruby community as a whole.

There's ZERO impact here on actual Listen usage. (I'm yet to proven wrong by a specific and compelling example).

> In a devops world, we should move fast and break things - including semver - but always retain our empathy for our users.

I want users to have the best quality software and experience. This includes: performance and memory benefits they get at little cost. This means migrating only once from 1.9.3 to 2.2 (or 2.3) instead of 1.9.3 -> 2.0 -> 2.1 -> 2.2 -> 2.3, which costs a LOT more time. Precious time.

Users aren't "all knowing" - so it's up to developers to encourage them to make the right choice.

Here, users can pin to Ruby 3.0.x. It's still a choice.

SemVer doesn't replace thinking and exceptional cases like this.

What's exceptional here is the ridiculously smooth migration path: Ruby 1.9.2 -> 2.0 -> 2.1 -> 2.2 -> 2.3 (1.9.2 -> 2.3). Doing so only brings benefits.

I don't understand how a dependency like that can cause any mentioned issue.

Up until now, no users were harmed during the bumping of just the minor version.

> especially when the cost is so low.

No, the cost is tremendous on the Ruby community. Just think about it: many users will unknowingly suffer from using an inferior version of Ruby for a whole friggin year! Think about all the contributors discouraged from still supporting the 1.8.7 syntax, because they've never seen actual 1.9.2 features used in the wild.

If you haven't noticed, the problem is no longer a lack of users - but not enough contributors to keep the software improving.

Too much dogma, policy and unreasonable "backward-compatibility".

If this is causing an actual (REAL LIFE) problem, someone please file an issue with the "actual" and "expected" described.

**Comment by rnelson0** [comment]
Before discussing this issue, I'd like to point out: I appreciate the Ruby lifecycle and want to see users move to the latest versions of it. I applaud all efforts to get development teams, operations, and especially vendors to the latest versions. However, I sit in Operations and so see the long tail of support surrounding Ruby and understand the compromises everyone involved has to make. I also think I was unclear in some of my statements, about cost in particular, and that confused the issue. I've attempted to clarify some of that below.

However, a deeper inspection shows the real problem lays in [guard's gemspec](https://github.com/guard/guard/blob/b2130a0efb51d30cc7a5878f0bd6dc7eeb1bb155/guard.gemspec#L20). This is also managed by the guard organization. It looks like there are a few options here:

1) Revoke listen 3.1.0, 3.1.1. Issue listen 3.next (3.0.7 or 3.1.2) stating that ruby < 2.2.0 is deprecated. Release listen 4.0.0 supporting only ruby >= 2.2.0.
2) Issue a new guard that pins to 3.0.6 and state that ruby < 2.2.0 is deprecated. Release guard 3.0.0 supporting only ruby >= 2.2.0.
3) Do nothing and force every user of ruby 1.9.3 + guard, which publicly advertises support for ruby >= 1.9.3, to work around this breakage independently, or stop using guard entirely.

Options 1 and 2 are realistic, easy, and work for everyone. Option 3 just leads more people to his over time or forces them away, neither of which are good long term results. If option 2 is the wisest course of action, I'd be happy to copy these arguments to that gem's issue tracker.

> How is Ruby 2.3 (with a minor version bump) not backward compatible with Ruby 2.0? (1.9.3 doesn't count, because it's EOL - people had way too long to upgrade already).

We (myself, @jyaworski, @bastelfreak) are not talking about the development of Ruby, we're talking about your gem specifically when we speak of backwards compatibility. When I use your gem with Ruby 1.9.3 and `>~ 3.0`, I do so because 3.0.6 supported Ruby 1.9.3. I simply do not expect that to change on a point release without warning (if there has been a deprecation warning already and I missed it, please let me know where it is!)

> Deprecate what exactly? "Ruby version" isn't an API or a functionality. It isn't an API I control within Listen.

An API specifies how software components interact, and your API previously stated that listen interacted with Ruby 1.9.3+. You have now deprecated support of `>= 1.9.3 and < 2.2.0`. This is why the gemspec has an option for Ruby versions the gem supports, because that's part of the API, and one that you explicitly control.

> Up until now, no users were harmed during the bumping of just the minor version.

Agreed. Unfortunately, your latest version harmed users - including your organization's own gem, guard.

> If this is causing an actual (REAL LIFE) problem, someone please file an issue with the "actual" and "expected" described.
Expected: When I use `gem "guard"`, I expect it to work with Ruby 1.9.3. This can be observed at https://travis-ci.org/voxpupuli/puppet-collectd/builds/124922164

Actual: When I use `gem "guard"` with Ruby 1.9.3, it no longer works. This can be observed at https://travis-ci.org/voxpupuli/puppet-collectd/builds/125300238

> No, the cost is tremendous on the Ruby community.

The cost I am referring to here is that of bumping the MAJOR version on listen (and/or guard). Is there a significant cost with letting `~>3.0` work with Ruby >= 1.9.3 and starting a `~> 4.0` train that only supports >= 2.2.0?

**Comment by e2** [comment]
> Release listen 4.0.0 supporting only ruby >= 2.2.0.

That's unreasonable. Try to get ANY gem owner to bump a major version JUST to allow Ruby 2.2-only PRs. Ask for Jekyll 4.0 release that's Ruby-2.2 only. (Won't happen. If it does, I'll reconsider).

> However, I sit in Operations

I don't want to be stereotypical, but only devops people are complaining here. I expected a valid use case (not related to provisioning, but actually something breaking in production).

> Issue a new guard that pins to 3.0.6 and state that ruby < 2.2.0 is deprecated. 

A lack of 2.2.0 support is a mistakenly "undocumented" de facto state in 3.x. The "uber-correct" solution would probably be to release 4.x just to deprecate the entire 3.x line, implement/backport Ruby 2.1 support (and earlier), and then release 5.x without the deprecation. But that's assuming Ruby versions aren't backward compatible, which is false.

> Do nothing and force every user of ruby 1.9.3 + guard, which publicly advertises support for ruby >= 1.9.3, to work around this breakage independently, or stop using guard entirely.

AFAIK, RubyGems simply won't install Listen 3.1 because the Ruby version doesn't match. If Rubygems doesn't resolve to Ruby 3.0.8 for Ruby 1.9.2, that's a RubyGems issue or "missing feature". Please open an issue there.

Also, Guard maybe supports 1.9.2 - I don't know. This needs to maybe be checked. You can lock to Listen 3.0.x and guard should work fine. So since no new release is needed to 'fix things', I don't see a problem the users can't resolve. It's up to the user to pin to Listen 3.x just to use an unsupported version of Ruby.

> We (myself, @jyaworski, @bastelfreak) are not talking about the development of Ruby, we're talking about your gem specifically when we speak of backwards compatibility. 

There is no breakage. Please show me an example of breakage. Since the user can lock any gem to any version they please (whether it makes sense or not), what you ask of is the users's freedom.

> When I use your gem with Ruby 1.9.3 and >~ 3.0, I do so because 3.0.6 supported Ruby 1.9.3.

That is false, because releases weren't tested for Ruby 1.9.3 compatibility. 

> if there has been a deprecation warning already and I missed it, please let me know where it is!

https://www.ruby-lang.org/en/news/2016/03/30/ruby-2-1-9-released/

No more bugfixes. Aka "no support". Security patches != support. That's because support means "fixing bugs".

It also shouldn't be Listen's responsibility to warn about Ruby being unsupported by Ruby-core developers. Just like it doesn't make every gem out there to tell you your Ruby version is no longer officially supported.

> An API specifies how software components interact, and your API previously stated that listen interacted with Ruby 1.9.3+. 

Which was probably wrong. I could've considered this just a bugfix and made a minor patch. I chose a minor version bump - so users had the freedom to avoid the patch if needed.

> Agreed.

I deeply respect you for that! So I can sleep well knowing no users were actually hurt. And any incompatibilities with 1.9.2 were correctly prevented by me bumping the version to the de facto Ruby 2.2 support.

> Unfortunately, your latest version harmed users - including your organization's own gem, guard.

How?

> Expected: When I use gem "guard", I expect it to work with Ruby 1.9.3.

"Work" vs "runs tests" - two different things. The problem is: you "expect" it to work, while it only "happens" to work. Your expectation is unreasonable to being with, because 1.9.3 is EOL.

```
Gem::InstallError: listen requires Ruby version ~> 2.2.
An error occurred while installing listen (3.1.1), and Bundler cannot continue.
Make sure that `gem install listen -v '3.1.1'` succeeds before bundling.
```

This is as expected. You need to Ruby ~> 2.2 to install this version of Listen.

What I find missing is bundler's lack of support to find a matching version (3.0.x) compatible with Ruby 1.9.3 (or whatever).

Solution for users: upgrade to Ruby 2.2 or don't update Listen. (Ask for backport releases instead?)

Solution for devops: lock to Listen 3.0.x

> Is there a significant cost with letting ~>3.0 work with Ruby >= 1.9.3 and starting a ~> 4.0 train that only supports >= 2.2.0?

Yes - first no one does that. (E.g. no Jekyll 4.x supporting only Ruby 2.3 where I can send PRs using Ruby 2.3-only features). Also, there's absolutely no API change, which is extremely misleading. API is the same, docs are the same.

But the real cost is: human time. Sure, discussing this issue is more exhausting than I would've imagined (given no users were actually affected by something that doesn't have a general+recommended solution).

"Devops" is not a valid use case in of itself. That's like calling "because programming" an excuse.

If you want a 1.9.2 branch for PRs, open an issue. if you want a 3.0.x release with those PRs, same thing - just ask.

I could do a deprecation release, but there's absolutely nothing to deprecate.

Whether it's a missing feature in Bundler to resolve this - that's a separate issue.

**Comment by rnelson0** [comment]
> I don't want to be stereotypical, but only devops people are complaining here. I expected a valid use case (not related to provisioning, but actually something breaking in production).

What about using `gem "guard"` is not a valid use case? And is testing not part of your production systems? They are here; if tests won't complete, our build pipeline rejects all changes. That's as production as it gets.

> You can lock to Listen 3.0.x and guard should work fine. So since no new release is needed to 'fix things', I don't see a problem the users can't resolve.

Anyone using guard, or guard-rake that requires guard, is affected because of the stated requirements in the guard gem. These three gems - listen, guard, and guard-rake - are all owned by the same organization. It's really weird that you're asking me to report an issue to Rubygems, that isn't the fault of rubygems, when its within the power of the organization to fix this just once for everyone affected. Instead, you are advocating distributed pain.

> That is false, because releases weren't tested for Ruby 1.9.3 compatibility. 

What is true is that listen was [advertising Ruby 1.9.3 compatibility](https://github.com/guard/listen/blob/41b2753417a8ac1133c8c90acea50182667849fc/listen.gemspec). It's not the user's fault that the authors were not testing it properly. 

> if there has been a deprecation warning already and I missed it, please let me know where it is!

Again, we're talking about *your* software here, not Ruby. Your software DOES NOT PUBLICLY DEPRECATE RUBY >= 1.9.3 and < 2.2.0 SUPPORT! I really do not understand why you keep referring everything upstream to Ruby when we're writing specifically in the context of listen.

> How?

Look at guard's gemspec. [Line 17](https://github.com/guard/guard/blob/b2130a0efb51d30cc7a5878f0bd6dc7eeb1bb155/guard.gemspec#L17) specifies that Ruby >= 1.9.3 is required. [Line 20](https://github.com/guard/guard/blob/b2130a0efb51d30cc7a5878f0bd6dc7eeb1bb155/guard.gemspec#L20) specifies that listen version >= 2.7 and < 4.0 will be included. *That gem is broken.* It has violated its API contract by no longer working with Ruby versions >= 1.9.3 and < 2.2.0.

> I deeply respect you for that! So I can sleep well knowing no users were actually hurt. And any incompatibilities with 1.9.2 were correctly prevented by me bumping the version to the de facto Ruby 2.2 support.

We have no agreement here. I don't even know where 1.9.2 came from, but I showed you demonstrable harm in the violation of the API contract and you are ignoring it.

> "Devops" is not a valid use case in of itself. That's like calling "because programming" an excuse.

You're the only one throwing that word around.

> Yes - first no one does that. (E.g. no Jekyll 4.x supporting only Ruby 2.3 where I can send PRs using Ruby 2.3-only features). Also, there's absolutely no API change, which is extremely misleading. API is the same, docs are the same.

Your API is your contract with your consumers and includes at least everything stated in the gemspec, such as support for using the gem with Ruby >= 1.9.3. When you change the version of dependencies for your gem, you have changed the API.

Furthermore, "everyone does this" or "no one does this" is not an excuse. At best, you're creating the same problems as everyone else.

Let's turn this around: show me the harm in bumping this to v4.0? I've shown you the harm to the guard and guard-rake gems, which promise Ruby 1.9.3 support, as thing stands. I'd like to see the harm caused by bumping the MAJOR version documented, so we can understand it.

**Comment by e2** [comment]
> What about using gem "guard" is not a valid use case? 

"listen requires Ruby version ~> 2.2." How is that a problem for users to understand?

I've updated the Readme just in case. Anyway, it's a RubyGems/Bundler issue that it doesn't resolve Listen to 3.0.8 (based on supported Ruby Version). There's no issue for this (yet).

> They are here; if tests won't complete, our build pipeline rejects all changes. 

Which is fantastic, because Listen 3.1 isn't deployed, and 3.0.x is still used in production.

Environment remains stable, despite dependency error. How is that not the goal of devops?

> It's really weird that you're asking me to report an issue to Rubygems, that isn't the fault of rubygems, when its within the power of the organization to fix this just once for everyone affected.

Because RubyGems is where the fix should be. Well, to be fair, RubyGems doesn't resolve versions - so it's "works as expected". Bundler resolves dependencies, but it doesn't take Ruby dependencies into account. (I don't know if it should - see https://github.com/bundler/bundler-features/issues/120).

I'm looking for a "clean fix" here, not an ugly workaround that doesn't address the root cause. All because I can create a workaround in Listen, doesn't mean I should.

UPDATE:  prerelease Bundler gracefully handles this: `gem install bundler --pre` and Bundler should correctly resolve to Listen 3.x.

> It's not the user's fault that the authors were not testing it properly.

Yes, it was my fault. There weren't any incompatibilities though, so I never noticed the issue.

> Your software DOES NOT PUBLICLY DEPRECATE RUBY >= 1.9.3 and < 2.2.0 SUPPORT!

It was mentioned in the README for a long time. I simply forgot to update the gemspec. (Ideally maybe the gemspec could extract the Ruby version from the readme?).

And again - there's nothing to deprecate. API is the same.

> I really do not understand why you keep referring everything upstream to Ruby when we're writing specifically in the context of listen.

You are referring to upstream Ruby. You are talking about incompatibilities which don't exists. You're going into ideology here about SemVer.

> Line 20 specifies that listen version >= 2.7 and < 4.0 will be included.

Oops! I just noticed: the dependency <=4 is wrong: https://github.com/guard/guard/issues/825

> It has violated its API contract by no longer working with Ruby versions >= 1.9.3 and < 2.2.0.

The following:

```ruby
s.required_ruby_version = ">= 1.9.3"
```

Doesn't not guarantee that Guard works with 1.9.3, because it cannot guarantee that dependencies will work with that version either.

That's the whole point of versioning: Guard only cares about Listen's API, not it's dependencies. If a dependency requires Ruby >= 2.x, it doesn't mean it has to blacklist all Listen versions except ones that support 1.9.3.

Dependencies are not a contract of "every supported version" - they're for SATISFYING.

> We have no agreement here. I don't even know where 1.9.2 came from, but I showed you demonstrable harm in the violation of the API contract and you are ignoring it.

I meant 1.9.3. Sorry.

It's not harm - because harm was prevented by unsatisfied dependencies. There's no API contract violated, because there's not API expectation violation here. Guard doesn't depend on a given Ruby API - it's very relaxed. Listen isn't. The overlap is Ruby 2.2. You can satisfy dependencies with no API incompatibility. Again everything works as advertised.

> You're the only one throwing that word around.

I'm just noticing a pattern that only devops people complain. And their argument is pretty much just that: devops. It's devops as opposed to "development" and "usage". Failed provisioning when dependencies are not satisfied is **expected behavior**. It's not "breakage" or harm.

I want specific cases, not being patronized constantly about "yes, we all enjoy new features, but you see - in devops ...". In understand devops. To the extent that I understand that a failure in provisioning due to unsatisfied dependencies is a good thing.

To make it clear: a failure in provisioning isn't "demonstrable harm". If you can't upgrade to the newest version of Apache due to failed dependencies, that's a good thing, right?

> Your API is your contract with your consumers and includes at least everything stated in the gemspec, such as support for using the gem with Ruby >= 1.9.3.

A range of versions doesn't imply every single version and patchlevel is supported. It all depends which dependent gems are available (e.g. all versions supporting 1.9.3. could've been yanked). It could've been Ruby 1.8.7, and it would've been fine until Listen supported 1.8.7 again (which might've broken things for real).

> When you change the version of dependencies for your gem, you have changed the API.

That doesn't make sense. I might've bumped a gem dependency because of a bug. That doesn't change the API an ANY level. And that's why a patchlevel version would've been sufficient.

> Furthermore, "everyone does this" or "no one does this" is not an excuse. At best, you're creating the same problems as everyone else.

When we talk about "expectations", it's not fair to have different expectations for the Listen gem than for other gems.

It's hypocrisy for you to be strict about the version in Listen, but relaxed when I give examples of doing the same in other gems.

I'm solving a core issue here by refusing to compromise without a compelling problem and a compelling solution indicating that changes in Listen are the way to go.

> Let's turn this around: show me the harm in bumping this to v4.0?

Few users will upgrade, and they'll be complaining and spamming with issues. And the fix will require EVERYONE to MODIFY their `Gemfile`. Or install/uninstall gems. Also, They won't be ABLE to update gems, because gem's can't support MULTIPLE MAJOR versions at the same time.

It would be the same as NEVER releasing 4.x. It would impractical. This means users won't get benefits of fixes or features.

It would make development a insane  "backward-oriented" effort. The death or Ruby if you extrapolate far enough. Imagine everyone stuck with Ruby 1.8.7 for 10 more years. How is that "no harm"?

The only "happy" people would be devops people. Everyone else would suffer.

> I've shown you the harm to the guard and guard-rake gems, which promise Ruby 1.9.3 support, as thing stands. I'd like to see the harm caused by bumping the MAJOR version documented, so we can understand it.

The major harm is: contributor discouragement. I don't get paid for backward compatibility support. It's no joy either.

The most important part: real users DON'T EVEN CARE. Their worst nightmare is "updating Ruby". And if that's a problem, that's not something Listen should even attempt to address.

**Comment by rnelson0** [comment]
On Monday, April 25, 2016, Cezary Baginski <notifications@github.com> wrote:

> What about using gem "guard" is not a valid use case?
>
> "listen requires Ruby version ~> 2.2." How is that a problem for users to
> understand?
>
That's Listen 3.1.0. Subscribers to Listen have had an API that stated Ruby
version >= 1.9.3 was required. You broke your own contract.

Ok, you say, but I changed it for listen! You changed your API then. You
argue that that's not the API. Unfortunately, guard and guard-rake both
state that they support Ruby 1.9.3 and listen < 4.0. This change literally
broke the dependency graph for guard. That is the root, underlying cause of
all other issues. You changed the API, which describes many aspects of how
the system interfaces with the application including the Ruby version it
supports, and broke your org's own gems' contract.

Everything else is minutiae.


-- 

Rob Nelson
rnelson0@gmail.com

**Comment by rnelson0** [comment]
> A range of versions doesn't imply every single version and patchlevel is
> supported.
>

Yes, actually, it does. It literally states "this gem depends on Ruby
versions of 1.9.3 or greater" or "this gem depends on the listen gem's
versions between X and Y." That's the entire point of stating your version
dependencies.

That's also why we are being affected, because we are relying on the API of
guard that says "this gem works with these Ruby versions and these
dependent gem versions" when that is demonstrably false. This is considered
harm, and again, the fix is simple. Mark your depredations in your API and
bump major versions. Not a big deal and it fixes the API contract
violations for all users, now and forever, in one place.


-- 

Rob Nelson
rnelson0@gmail.com

**Comment by e2** [comment]
> Subscribers to Listen have had an API that stated Ruby version >= 1.9.3 was required. 

version => 1.9.3 doesn't imply that 1.9.3 is supported.

It's a version **constraint** not a version **contract**.

**Comment by e2** [comment]
> This change literally broke the dependency graph for guard.

There's no such thing as a "broken dependency graph". Either you can satisfy dependencies, or you can't. Both case are valid.

**Comment by e2** [comment]
> You changed the API, which describes many aspects of how the system interfaces with the application including the Ruby version it supports, and broke your org's own gems' contract.

Please show me Ruby code that produces an incorrect result due to the API being changed. In this case, no Ruby code is executed against the API, but you can't even install it.

**Comment by e2** [comment]
> That's also why we are being affected, because we are relying on the API of guard that says "this gem works with these Ruby versions and these dependent gem versions" when that is demonstrably false. 

Dependencies are not part of the API. What Ruby provides in a backward-compatible way is not Listen's responsibility to manage and declare. 

Ruby 2.2 features are not exposed **through** Listen's API, they're exposed **on top of** Listen's API. It's up to Guard to decide whether to use those features or not.

If dependencies WERE part of the API, then supporting every next Ruby patchlevel or minor would require a version bump.Yes, you can do that, but it isn't necessary!

I mean, you can take it to the extreme and make EVERY change a MAJOR bump!

A patchlevel basically means: "bugfix - no API change".
A minor means: "new methods, arguments added to API".
A major means: "you likely have to change the consumer code for things to work".

Here, Guard doesn't need ANY changes to work with Listen's API from 3.1.0. Dependencies or not. As long as you satisfy Listen's new backward-compatible dependency on Ruby 2.2.

So even if you somehow treat dependencies as an API, those dependencies are backward compatible. (Listens changes in the dependency on Ruby 2.2. doesn't require changes in Listen API usage in Guard or even in the dependencies in Guard).

And since your "dependency API" is backward compatible (meaning: no changes necessary in Guard to satisfy dependencies) ... there's no need for a major bump!

**Comment by e2** [comment]
> Mark your depredations in your API and bump major versions.

If Ruby deprecates something, why should Listen care and declare that? It's outside the scope/responsibility of Listen.

And since Listen shouldn't care about deprecated features in Ruby ... it means it shouldn't care at all what features Ruby provides or not, as long as they're backward-compatible to the extent Listen uses it internally.

You can think about it this way: if Guard can marshal data with Ruby 2.1, and then Listen is updated, and then Guard unmarshals that data with Ruby 2.2 (because Listen bumped the dependency), then compatibility is preserved. So: no major bump needed. 

(Of course, the mistake was still having the constraint relaxed to Ruby 1.9.3 in the Listen gemspec, while the Readme stated that 1.9.3 was no longer supported. So you were accidentally using an unsupported version of Ruby that happened to work. So you were simply lucky and nothing more).

**Comment by e2** [comment]
Summary (since I've responded to all the points): 

The misunderstanding comes from Listen working on 1.9.3, when it officially wasn't supported. This became an unfortunate expectation and that expectation was broken. But it was broken on the dependency resolution level, so this is a normal case of SemVer doing what it does.

Fictional example: I could've merged a PR with Ruby 2.x - only changes. According to the Readme (which is covered by SemVer too), this would've been ok. According to the gemspec, the constraint of including 1.9.3 was just too relaxed.

All because 3.0.8 works on 1.9.3, that doesn't mean it was expected to.

So locking to 3.0.8 just perpetuates the problem.

**Comment by rnelson0** [comment]
> All because 3.0.8 works on 1.9.3, that doesn't mean it was expected to.
>
As a user, when I see a gem that states the Ruby version in the gem spec is
>= 1.9.3, I expect that it works with Ruby down to 1.9.3.


-- 

Rob Nelson
rnelson0@gmail.com

**Comment by e2** [comment]
@rnelson0 - a Travis build matrix is not the same as a version constraint. A Travis build matrix gives you the false assumption that a given Ruby is supported by a set of gems.

Analogy: redcarpet doesn't work with JRuby (AFAIR), because of compiled extensions which JRuby doesn't support. So, if the redcarpet gem does not restrict it's Ruby version requirement, it still doesn't mean it's expected to work on JRuby.

So all because you put Ruby 1.9.3 inside a `.travis.yml`, that doesn't mean anything is automatically expected to work, regardless of how relaxed the version constraint is. Some gems have no version constraint, because when they were last released, Ruby 1.8.7 was the last available Ruby version.

Broken travis build != SemVer violation.

**Comment by rnelson0** [comment]
I'm not talking about travis.yml. I'm talking about this line, [right here](
https://github.com/guard/listen/blob/41b2753417a8ac1133c8c90acea50182667849fc/listen.gemspec#L25).
Nothing else.

**Comment by rymai** [comment]
I just want to thank @e2 for pushing the adoption of Ruby 2.2+ forward!

I find it really good that **installation** fails, not **usage**.

What's even more awesome is that new Bundler will bring the best of both worlds: smooth installation and early dependency on new Ruby versions in `.gemspec` files!

**Comment by rnelson0** [comment]
@rymai No-one has said, don't support Ruby 2.2. We're just asking for semver-compatible numbering of listen, so that those using ruby 1.9.3 don't have breakage in their CI/CD pipeline (where fresh installations are common to avoid untracked dependency issues) because of upstream gem dependencies.

**Comment by e2** [comment]
@rnelson0 - if there's one thing you don't understand here, it's that a version constraint doesn't imply a given version is supported. The only time when your statement would be valid would be if guard was LOCKED to Ruby 1.9.3. But even then, it's not incorrect to have a library that can no longer be installed.

`>= 1.9.3` only means it's POSSIBLE that 1.9.3 could work.

POSSIBLE != MUST SUPPORT

There is no mention of "must support all matching versions" in the SemVer.

It also doesn't mean I have to support AND TEST EVERY SINGLE DARN PATCHLEVEL OF RUBY all because I have "=>" in the gemspec.

There's a grey area: there are versions expected to work, there are versions expected not to work ... and there are versions where it's UNKNOWN whether it works. That's the case with 1.9.3. It's UNKNOWN whether it works, which means it's UNSUITABLE for devops/production.

And Travis matters, because it's the best place to describe which versions are supported. If it isn't tested by Travis, how can you expect it to be supported?

Example: https://travis-ci.org/guard/guard

1.9.3 isn't tested by Travis. If you rely on tests for compatibility, why don't you rely on `.travis.yml` or the README?

As a solution, I created this: https://github.com/guard/listen/pull/376

Now the version constraint is created automatically by what's in `.travis.yml`.

Single responsibility principle.

So now, if someone wants Ruby 2.3 or 2.4 support, all that needs to be done is a change in `.travis.yml` and a patchlevel version bump (if a version was added) or a minor version bump (if a backward-compatible version was removed or a backward-incompatible version of Ruby was removed). This is because is should be up to the package manager or version resolver to match the right supported version.

**Comment by rnelson0** [comment]
> @rnelson0 - if there's one thing you don't understand here, it's that a version constraint doesn't imply a given version is supported. The only time when your statement would be valid would be if guard was LOCKED to Ruby 1.9.3. But even then, it's not incorrect to have a library that can no longer be installed.

http://guides.rubygems.org/specification-reference/#required_ruby_version says otherwise about your use of the stanza. That's not part of SemVer, but part of rubygems' gemspec specifications.

> And Travis matters, because it's the best place to describe which versions are supported. If it isn't tested by Travis, how can you expect it to be supported?

Because your gemspec said it only requires ruby 1.9.3 or greater.

**Comment by e2** [comment]
@rnelson0 - you're not understanding the term "requires" in this context.

It means: "this gem is declared to REQUIRE RUBY at given version". 

It does not mean "this gem is REQUIRED to SUPPORT RUBY AT THIS VERSION".

Look, the only way RubyGems can "enforce" this is: by not installing the gem.

That's the ONLY action `gem install` can take: NO ACTION when the version doesn't match.

So practically, ">= 1.9.3" means: "anything other than => 1.9.3 will FAIL TO INSTALL". It doesn't say: "1.9.3 MUST WORK because it's included by the constraint in the gemspec".

All because something CAN BE INSTALLED doesn't imply IT IS EXPECTED TO WORK.

Again: here, THE CURRENT RUBY IS REQUIRED TO BE at some version.

This does NOT mean Listen is REQUIRED to "be" or "do" anything.

Listen is setting the requirement on the Ruby version. It doesn't mean the Ruby version is setting a requirement on Listen.

It's not a bi-directed graph. The requirement doesn't go both ways.

If it was, THAT would be a broken dependency (like you mentioned before).

**Comment by rnelson0** [comment]
Of course it's a dependency. Otherwise you wouldn't have updated it to
exclude versions below 2.2.0, or even provided it at all (it is optional).

**Comment by e2** [comment]
> Otherwise you wouldn't have updated it to exclude versions below 2.2.0, or even provided it at all (it is optional).

I provided it to inform users that it's no longer "might work" but now it's "don't expect it to work".

The word "dependency" can mean many things. You've narrowed it down to "must support Ruby 1.9.3 because I can install it and it worked previously". You see a guarantee where there was no guarantee in the first place.

You've created a false assumption that you believed in. This change prevents you from making that false assumption again. It's your assumption that is affected by this change. Not Semver or RubyGems spec.

> Otherwise you wouldn't have updated it to exclude versions below 2.2.0, or even provided it at all (it is optional).

Yes, Listen depended on Ruby >= 1.9.3 to install. But the reverse isn't true: that Ruby 1.9.3 requires listen to be possible to install.

> or even provided it at all (it is optional).

Yes - and what does optional mean? What does it mean if the required_ruby_version is omitted? Does that mean that ALL POSSIBLE RUBY VERSIONS are supported? Or MUST be supported?

Of course not. No "contract" here.

**Comment by e2** [comment]
If you allow me to summarize the misunderstanding:

As a devops-aware person, you've made an assumption much stricter than SemVer (or RubyGems spec) defines.

And because of your assumption, you're treating a perfectly valid use case (gem not installing because of missing dependencies) as breaking SemVer or a invalid dependency specification.

Your assumption is way too strict, because it shuts off a perfectly valid use case. This issue doesn't affect users at all. There is a clean solution outside Listen for this.

Your assumption has no valid use case for users that Listen "must fix".

Your overly strict assumption simply excludes other valid possibilities. SemVer simply as strict as you'd like it to be.

If you like, you can propose an alternative to SemVer. But since you're the one limiting the scope of "what's valid", you're creating a heavy burden with no benefit.

Summary: your assumption about SemVer/RubyGems was incorrect in this use case.

If I bumped it to 4.x, users would still be using a misleading version constraint, and they wouldn't find out until upgrading to 4.x. That's bad for users, wouldn't you agree?

**Comment by e2** [comment]
In case this is a better explanation: you were in a DANGEROUS situation.

Before Listen 3.1, you assumed 1.9.3 worked! And by SHEER LUCK it did.

Isn't that ALREADY a dangerous devops situation? You assumed a version was supported and it wasn't!

You should be HAPPY that now you know that what you thought was supported was not!

You should be HAPPY that things "broke" not in production (by e.g. another 3.0.x patch using a Ruby 2.x -only feature) but during tests!

In short: I potentially saved your ass here by NOT bumping a major version!

**Comment by rnelson0** [comment]
On Tue, Apr 26, 2016 at 9:32 AM, Cezary Baginski <notifications@github.com>
wrote:

> Before Listen 3.1, you assumed 1.9.3 worked! And by SHEER LUCK it did.


I assumed it, because your gemspec said it did and there were no failures
in using it in test.

> In short: I potentially saved your ass here by NOT bumping a major
version!
That's an odd perspective. This change caused nothing but heartache for
some users of your gem. I don't feel saved.

**Comment by e2** [comment]
@rnelson0 - this is getting nowhere. You're hell-bent on forcing me to agree that a version constraint means a version is fully supported in production.

My gemspec did "say" anything. It didn't match the version declared in the README and it didn't match the tested versions on Travis. So you basically used an UNTESTED combination of Ruby and Listen.

And you are unhappy that the UNTESTED combination stopped working. That's the **opposite** approach of devops.

With my commit, that's no longer possible for every 3.1.x version from now on.

> That's an odd perspective.

Your perspective is odd, because it isn't based on any practical usage. There isn't any documented behavior (SemVer) or specific reasoning behind it.

Also, Ruby is an interpreter and not a compiler. You need to take that into account.

> This change caused nothing but heartache for some users of your gem.

That's just speculation.

> I don't feel saved.

You should. Even if you don't, a version constraint's goal isn't to "soothe feelings" but avoid technical problems. 

**Currently, if you stick to Listen 3.0.x, and someone accepts a Ruby 2.x-only patch into a 3.0.x version of Listen, things will break for you in PRODUCTION!**

Ask a 3rd party to read this thread and decide whether your expectations are reasonable or not. That's because you've decided I'm wrong, and all you're doing is trying to find a way to prove me wrong.

To prevent this in the future, I created this gem: https://github.com/e2/ruby_dep

This will make sure the declared supported Ruby version will match what is tested on Travis.

And no matter how you feel, I'm assuming you agree that bumping Listen at this point makes no sense, no matter how you "interpret" the what a ruby constraint in a gemspec means.

(Because you'll still have a version of Listen that claims to support Ruby 1.9.3, but actually officially doesn't).

So unless you have strong arguments to convince me that your assumptions are correct and practical, I suggest we end this.

To focus on the solution: you haven't given me strong, compelling, specific, smart reason for a major bump.

**Comment by e2** [comment]
One more thing: https://www.ruby-lang.org/en/news/2015/12/16/unsafe-tainted-string-usage-in-fiddle-and-dl-cve-2015-7551/

Where it states:

**"Affected Versions: All patch releases of Ruby 1.9.2 and Ruby 1.9.3 (DL and Fiddle)."**

This is since December 2015.

**Comment by e2** [comment]
I've addressed the points, and the decision for Listen is: 

- changes in required Rubies do not need a major version bump (Ruby version is not part of an API Listen has enough control over)
- this is valid according to SemVer
- only supported Rubies will be supported, unless backports are requested.
- no actual issues affecting end users (broken Travis builds don't count)
- locking to 3.0.x is neither a solution, nor recommended (instead, upgrade Ruby)
- further discussion is beyond the scope of Listen (if not, open a new issue)

For a more general discussion about Ruby support, I suggest: https://gist.github.com/e2/ac32569852cbd31f7da637500174d907

Please comment there on any specific points.

Thank you.
